File: backend/models/map.go
package models

import (
	"github.com/google/uuid"
)

type GameMap struct {
	ID            uuid.UUID  `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	OwnerID       uuid.UUID  `gorm:"type:uuid;not null" json:"owner_id"`
	RoomCode      string     `gorm:"type:varchar(50);unique;not null" json:"room_code"`
	Name          string     `gorm:"type:varchar(255);not null" json:"name"`
	BackgroundURL string     `gorm:"type:text" json:"background_url"`
	GridRows      int        `gorm:"default:20" json:"grid_rows"`
	GridCols      int        `gorm:"default:20" json:"grid_cols"`
	TileSize      int        `gorm:"default:50" json:"tile_size"`
	Tokens        []MapToken `gorm:"foreignKey:MapID;constraint:OnDelete:CASCADE" json:"tokens"`
}

type MapToken struct {
	ID             uuid.UUID  `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	MapID          uuid.UUID  `gorm:"type:uuid;not null" json:"map_id"`
	CharacterID    *uuid.UUID `gorm:"type:uuid" json:"character_id"`
	AssignedUserID *uuid.UUID `gorm:"type:uuid" json:"assigned_user_id"`
	Name           string     `gorm:"type:varchar(255);not null" json:"name"`
	ImageURL       string     `gorm:"type:text" json:"image_url"`
	TokenType      string     `gorm:"type:varchar(50);default:'npc'" json:"token_type"` // "pc" or "npc"
	GridX          int        `gorm:"not null" json:"grid_x"`
	GridY          int        `gorm:"not null" json:"grid_y"`
	Size           int        `gorm:"default:1" json:"size"`
	Color          string     `gorm:"type:varchar(50)" json:"color"`
	Visible        bool       `gorm:"default:true" json:"visible"`
	InitiativeOrder *int      `gorm:"type:int" json:"initiative_order,omitempty"`
}

File: frontend/src/types/map.ts
/** Battle map token role (matches backend models.MapTokenType). */
export type MapTokenType = 'pc' | 'npc';

export interface MapToken {
  id: string;
  map_id: string;
  character_id?: string;
  monster_id?: string;
  assigned_user_id?: string;
  name: string;
  image_url: string;
  token_type: MapTokenType;
  grid_x: number;
  grid_y: number;
  size: number;
  color: string;
  visible: boolean;
  initiative_order?: number;
}

export enum MapElementType {
  Wall = "wall",
  DifficultTerrain = "difficult_terrain",
  Trap = "trap",
  Elevation = "elevation",
}

export interface TrapProperties {
  id: string;
  map_element_id: string;
  name: string;
  description: string;
  visible_to_players: boolean;
}

export interface DifficultTerrainProperties {
  id: string;
  map_element_id: string;
  color: string;
}

export interface ElevationProperties {
  id: string;
  map_element_id: string;
  level: number;
}

export interface WallProperties {
  id: string;
  map_element_id: string;
  details: string;
}

export interface MapElement {
  id: string;
  map_id: string;
  type: MapElementType;
  grid_x: number;
  grid_y: number;
  trap_properties?: TrapProperties;
  difficult_terrain_properties?: DifficultTerrainProperties;
  elevation_properties?: ElevationProperties;
  wall_properties?: WallProperties;
}

export interface GameMap {
  id: string;
  owner_id: string;
  room_code: string;
  name: string;
  background_url: string;
  grid_rows: number;
  grid_cols: number;
  tile_size: number;
  tokens: MapToken[];
  elements: MapElement[]; // New field
}

export interface CreateMapRequest {
  name: string;
  room_code: string;
  background_url: string;
  grid_rows: number;
  grid_cols: number;
  tile_size: number;
}

export interface UpdateMapRequest {
  name?: string;
  background_url?: string;
  grid_rows?: number;
  grid_cols?: number;
  tile_size?: number;
}

export interface CreateTokenRequest {
  character_id?: string;
  monster_id?: string;
  assigned_user_id?: string;
  name: string;
  image_url: string;
  token_type: MapTokenType;
  grid_x: number;
  grid_y: number;
  size: number;
  color: string;
  visible: boolean;
}

export interface CreateMapElementRequest {
  type: MapElementType;
  grid_x: number;
  grid_y: number;
  trap_name?: string;
  trap_description?: string;
  trap_visible_to_players?: boolean;
  difficult_terrain_color?: string;
  elevation_level?: number;
  wall_details?: string;
}

export interface UpdateMapElementRequest {
  type?: MapElementType;
  grid_x?: number;
  grid_y?: number;
  trap_name?: string;
  trap_description?: string;
  trap_visible_to_players?: boolean;
  difficult_terrain_color?: string;
  elevation_level?: number;
  wall_details?: string;
}

export interface UpdateTokenRequest {
  grid_x?: number;
  grid_y?: number;
  visible?: boolean;
  size?: number;
  color?: string;
  assigned_user_id?: string;
  character_id?: string | null;
  monster_id?: string | null;
  name?: string;
  image_url?: string;
  token_type?: MapTokenType;
}
